gradido/gradido
# Blockchain-Connector openCommunication

## 🍰 Pullrequest
First Pull-Request for connecting the Apollo-based backend with Blockchain-Connector

The last requirement from moriz was that the communication between Community-Server (Apollo backend)
and Blockchain-Connector should start like the communication between Communities with a openCommunication request.
[CommunityCommunication in Concepts in gradido](https://github.com/gradido/gradido/blob/master/docu/Concepts/TechnicalRequirements/CommunityCommunication.md)

With that to get a simple transfer transaction onto blockchain this requests are needed:

![activity-diagram](https://gradido.github.io/gradido_blockchain_connector/uml_open_connection_communication1.png)


In this Pull-Request only the first request is implemented alongside some boilerplate stuff. 


### Boilerplate stuff
```.env
COMMUNITY_PUBLIC_KEY=f7bf12696ca09c7a348dc48c82894eacbfb955029125f28fd226510def83d53d
COMMUNITY_PRIVATE_KEY=09e552a127f999cc2d339d220f6afa6a31cfbcbcf2c25f10ff9af09c29b49dac
[...]
# Blockchain Connector
BLOCKCHAIN_CONNECTOR_PUBLIC_KEY=d930443eab996244813dd184105b58d490ae131cfb5ed464768939261031b875
BLOCKCHAIN_CONNECTOR_API_URL=http://blockchain-connector:1271
```

This was added to backend .env (and therefore .env.dist), with that `CONFIG_VERSION` is changed to: `v10.2022-09-19`
`COMMUNITY_PUBLIC_KEY` and `COMMUNITY_PRIVATE_KEY` can be generated with hyperswarm: 
```ts
import DHT from '@hyperswarm/dht'

const keyPair = DHT.keyPair()
```

but the hyperswarm import is waiting in #1874 so I decide to better not mess with that and write the keys direct into config. 

`BLOCKCHAIN_CONNECTOR_API_URL` is needed for connecting with Blockchain-Connector via JSON-Rest API. 
To make the test easy a Blockchain-Connector Build is waiting as Docker-Image, updated with the ability to get the config parameter via environment
and it can use also sqlite db so no need for setting up a database for blockchain connector.
To use this, this was added to docker-compose.override.yml: 
```yml
#########################################################
## BLOCKCHAIN CONNECTOR #################################
#########################################################
blockchain-connector:
  image: gradido/blockchain_connector:0.1
  networks: 
  - internal-net
  - external-net
  environment:
    - DB_SQLITE=db.sqlite
    - CRYPTO_JWT_PRIVATE=d7c682d1bc9ef8ab35176e8bd5644a088cc7ab7c3c03c5fe40a4c58db8a3e4c2d930443eab996244813dd184105b58d490ae131cfb5ed464768939261031b875
    - JSON_SERVER_PORT=1271
  volumes:
  -  ./logs/blockchain_connector:/var/log/gradido
```
- `DB_SQLITE`: with setting this to a file name sqlite will be used instead of mysql
- `CRYPTO_JWT_PRIVATE`: here goes the ed25519 private key for the blockchain connector

For the full list of options look here: 
[Dockerfile](https://github.com/gradido/gradido_blockchain_connector/blob/master/Dockerfiles/Dockerfile.release) or
[blockchain_connector.properties](https://github.com/gradido/gradido_blockchain_connector/blob/master/config/blockchain_connector.properties)

`BLOCKCHAIN_CONNECTOR_PUBLIC_KEY` is the corresponding public key for the `CRYPTO_JWT_PRIVATE` from Blockchain-Connector.

Note: The whole ed25519 secret (private) key written out contains the public key after the half, but for Community Key I have decided to store
only the unique part of the secret key, because when used it can easy be concatenated.

By pushing to blockchain-connector master a new docker image will be generated with the tag: `gradido/blockchain_connector:master`,
`gradido/blockchain_connector:0.1` is manually uploaded to have a fixed version for this Pull-Request.

### Why was community-key-B changed to signature?
I had said that it is possible to encrypt something with a private key so that all knowing the public key can decrypt it.
Sadly that's not quite right, it's only the way it is stored in my memory. 
It exist a way to generate a shared secret with can be computed from Public Key A and Private Key B or Public Key B and Private Key A. 
With that a symmetric encryption is possible so that only A and B can decrypt the message. 
But for this x25519 keys are needed, so ed25519 keys must be transformed first. 
And also a nonce is recommended to prevent guessing the keys if someone get's hand on many encrypted messages.
So to encrypt something with that way 24 Bytes for the nonce is added and 16 Bytes for MAC-Bytes which are needed. 
When also community-key-B should be encrypted with that we get as result 72 Bytes encrypted message, or 144 in hex format. 

But I prefer the simpler way (both in CPU and bandwidth consumption) with a signature. 
With ed25519 keys arbitrary messages can be signed with the private key producing a 64 Bytes long Signature, or 128 in hex format,
which can be verified with the public key. 
In this Scenario: [CommunityCommunication in Concepts in gradido](https://github.com/gradido/gradido/blob/master/docu/Concepts/TechnicalRequirements/CommunityCommunication.md) the receiver already knows the public key B because it is his own, so we don't need to transmit it.
We only need the signature. For verify Community B or the Blockchain Connector simply puts his own Public Key in the function, the public Key from A and the signature. So he can checks that the sender must posses the private key A. 
What is also possible, that the sender was able to intercept the communication between A and B and has simply stolen the signature and can pretend to be community A. 


### Testsetup
To test that the openCommunication call works together with the Blockchain-Connector one must do:
- copy `COMMUNITY_PUBLIC_KEY`, `COMMUNITY_PRIVATE_KEY`, `BLOCKCHAIN_CONNECTOR_PUBLIC_KEY`, `BLOCKCHAIN_CONNECTOR_API_URL` and `CONFIG_VERSION` from `.env.dist` to you actually .env file
- call `docker-compose build && docker-compose up` and lookout for `blockchain-connector_1  |  call /openCommunication` and `jwt token from blockchain connector received:` when it work, it should be shown a serialized jwt token.

If it don't work the Blockchain-Connector logfiles can be found in `logs/blockchain_connector`

### Code-Changes
The code for the openCommunication request was added in the file CommunityCommunication.ts in the new folder federation. 
I think the federation stuff goes into this folder, or exist other plans for this?

For test the openCommunication function it is called in createServer.ts on startup, but I will remove this code before merging.
For me this was the easiest way to test this, but I'm open for suggestions how to test parts for a process which isn't finished yet.
Of course it is possible to write tests for that and mock the blockchain-connector api but then whe still don't know if this work together.

It is also possible to update the test workflow that a real blockchain-connector is used for testing but I do not think you like this idea. 

UPDATE: 
removed code in createServer because it crash the tests. 
to try it out yourself you need to checkout commit [7dcfc15](https://github.com/gradido/gradido/pull/2230/commits/7dcfc15fae0365c895ab27b4d09c2405ad991ac0)
